core/parentchain/indirect-calls-executor/src/filter_metadata.rs

```diff
@@ -17,35 +17,80 @@
 
 use crate::{
 	error::Result,
+	event_filter::{FilterEvents, MockEvents},
 	indirect_calls::{CallWorkerArgs, ShiedFundsArgs},
-	parentchain_extrinsic_parser::ParseExtrinsic,
+	parentchain_parser::ParseExtrinsic,
 	IndirectDispatch, IndirectExecutor,
 };
 use codec::{Decode, Encode};
 use core::marker::PhantomData;
+use itp_api_client_types::{Events, Metadata};
 use itp_node_api::metadata::{NodeMetadata, NodeMetadataTrait};
+use itp_types::H256;
 
-/// Trait to filter an indirect call and decode into it, where the decoding
-/// is based on the metadata provided.
-pub trait FilterCalls<NodeMetadata> {
-	/// Call enum we try to decode into.
-	type Call;
+pub trait EventsFromMetadata<NodeMetadata> {
+	type Output: FilterEvents;
+
+	fn create_from_metadata(
+		metadata: &NodeMetadata,
+		block_hash: H256,
+		events: &[u8],
+	) -> Option<Self::Output>;
+}
 
-	/// Knows how to parse the parentchain extrinsics.
-	type ParseParentchainExtrinsic;
+pub struct EventCreator;
+
+impl<NodeMetadata: TryInto<Metadata> + Clone> EventsFromMetadata<NodeMetadata> for EventCreator {
+	type Output = Events<H256>;
+
+	fn create_from_metadata(
+		metadata: &NodeMetadata,
+		block_hash: H256,
+		events: &[u8],
+	) -> Option<Self::Output> {
+		let raw_metadata: Metadata = metadata.clone().try_into().ok()?;
```

Comment:
To pass a reference and then call clone to get an owned value is an anti-pattern.